File: benchmarks/smith_waterman/cpp/mem.cpp
// Measures peak heap allocation of sw_align at the largest benchmark input
// size (5_000 bp). Prints a single JSON line to stdout.

#include "../../common/cpp/mem_track.hpp"

#include <cstdio>
#include <fstream>
#include <stdexcept>
#include <string>

#include "sw.cpp"

static std::string load_file(const std::string& path)
{
    std::ifstream ifs(path, std::ios::binary);
    if (!ifs)
        throw std::runtime_error("Cannot open " + path +
            " - run benchmarks/data/prepare_data.py first");
    std::string s((std::istreambuf_iterator<char>(ifs)), {});
    while (!s.empty() && (s.back() == '\n' || s.back() == '\r'))
        s.pop_back();
    return s;
}

int main()
{
    const auto query   = load_file("../../data/prepared/seq_pair_a.txt");
    const auto subject = load_file("../../data/prepared/seq_pair_b.txt");

    mem_track::reset();
    auto result = thesis_sw::sw_align(query, subject);
    volatile uint16_t sink = result.sw_score;
    (void)sink;
    const std::size_t peak = mem_track::peak_bytes();

    std::printf("{\"peak_bytes\": %zu}\n", peak);
    return 0;
}

File: benchmarks/smith_waterman/rust/examples/mem.rs
// measures peak heap allocation of sw_align_default at n=5_000.
// prints {"peak_bytes": <N>} to stdout.

#[global_allocator]
static ALLOC: dhat::Alloc = dhat::Alloc;

use std::hint::black_box;
use std::path::Path;

fn data_path(file: &str) -> std::path::PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("../../data/prepared")
        .join(file)
}

fn main() {
    let seq_a = std::fs::read(data_path("seq_pair_a.txt"))
        .expect("seq_pair_a.txt not found - run benchmarks/data/prepare_data.py first");
    let seq_b = std::fs::read(data_path("seq_pair_b.txt"))
        .expect("seq_pair_b.txt not found - run benchmarks/data/prepare_data.py first");

    let _profiler = dhat::Profiler::builder().testing().build();

    black_box(smith_waterman::sw_align_default(black_box(&seq_a), black_box(&seq_b)));

    let stats = dhat::HeapStats::get();
    println!("{{\"peak_bytes\": {}}}", stats.max_bytes);
}
